streams: Add new can_unsubscribe_group permission setting.

Aditya8840 opened the pull request:
This PR introduces a new channel setting, 'Who can unsubscribe from this channel?' ensuring that organization admins always retain the ability to unsubscribe themselves.

Fixes: #33413.

[#design > Self-Unsubscribe Setting @ 💬](https://chat.zulip.org/#narrow/channel/101-design/topic/Self-Unsubscribe.20Setting/near/2132009)
[#api design > Self-Unsubscribe Setting @ 💬](https://chat.zulip.org/#narrow/channel/378-api-design/topic/Self-Unsubscribe.20Setting/near/2126895)


**Screenshots and screen captures:**
| Channel Settings                  | Channel Popover                  | Advanced configurations                  |Channel Settings (Left panel)                  | Org-Level Permission                |
|-----------------------------------|----------------------------------|----------------------------------|----------------------------------|----------------------------------|
| ![](https://github.com/user-attachments/assets/e17846dd-190d-4f2c-9d20-2a2d80ae2513) | ![](https://github.com/user-attachments/assets/f1ed5da3-84ce-4a12-9e3c-188cd785dafb) | ![](https://github.com/user-attachments/assets/18cd94f1-2c70-421b-81d1-35f67de1d881) | ![](https://github.com/user-attachments/assets/2d0aaf21-d6f5-477f-adc5-1121a5b48a4b) | ![](https://github.com/user-attachments/assets/972b7943-2345-4c67-8880-f34b8aa98534) |


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

Aditya8840 committed:
streams: Add new can_unsubscribe_group permission setting.

Fixes: #33413.

Aditya8840 commented on zerver/lib/streams.py:
```diff
@@ -1290,7 +1290,7 @@ def list_to_streams(
     user_profile: UserProfile,
     autocreate: bool = False,
-    unsubscribing_others: bool = False,
+    unsubscribing_others: bool | None = None,
```
Should I change the name of this parameter since its purpose has changed? It now checks whether it’s called from an unsubscribe action, for either self-unsubscribe or unsubscribing others.

Aditya8840 commented:
@alya , Ready for your review!

alya commented:
Let's just label it "Who can unsubscribe from this channel".

Aditya8840 commented:
> Let's just label it "Who can unsubscribe from this channel".

Done!

sahil839 commented on zerver/openapi/zulip.yaml:
```diff
@@ -21051,6 +21076,26 @@ paths:
+
+                        [update-group-setting]: /api/group-setting-values#updating-group-setting-values
+
+                        Note that a user must [have content access](/help/channel-permissions)
```
I think this is incorrect as per the code now. User with metadata access and permission to administer can change this setting. And this is also inconsistent with the description below for the same setting.

sahil839 commented on web/src/stream_data.ts:
```diff
@@ -611,14 +611,26 @@ function can_administer_channel(sub: StreamSubscription): boolean {
+        return true;
+    }
+
+    return settings_data.user_has_permission_for_group_setting(
```
This should also be changed as per the permissions change in backend.

sahil839 commented on web/src/stream_events.ts:
```diff
@@ -104,7 +104,11 @@ export function update_property<P extends keyof UpdatableStreamProperties>(
+        if (
+            property === "can_subscribe_group" ||
+            property === "can_add_subscribers_group" ||
+            property === "can_unsubscribe_group"
```
This will also need `can_unsubscribe_others_group`.

sahil839 reviewed (commented):
Posted some comments. Haven't tested manually yet since this might need some permissions change.

Aditya8840 commented:
Updated the PR, @sahil839. It’s ready for review again!

Aditya8840 commented on zerver/views/streams.py:
```diff
@@ -546,11 +548,13 @@ def remove_subscriptions_backend(
 
     else:
         people_to_unsub = {user_profile}
+        unsubscribing_self = True
```
Should `unsubscribing_self` be set to `True` when the target is their own bot?

sahil839 replied:
I guess it should be, but we should discuss it once on CZO.

Aditya8840 replied:
Updated the PR, @sahil839. It’s ready for review again!

Aditya8840 committed:
streams: Separate bulk_check_basic_stream_access function.

This commit extracts the common logic for bulk-checking basic stream
access into a separate function, reducing redundancy when introducing
can_unsubscribe_group.

sahil839 commented on zerver/lib/streams.py:
```diff
@@ -1114,6 +1122,37 @@ def bulk_can_remove_subscribers_from_streams(
+        get_recursive_membership_groups(user_profile).values_list("id", flat=True)
+    )
+
+    permission_failure_streams: bool = True
```
Since we do not need to store streams here in this variable, we can use some other name `cannot_administer_atleast_one_stream`.

sahil839 commented on web/src/stream_data.ts:
```diff
@@ -619,14 +619,28 @@ function can_administer_channel(sub: StreamSubscription): boolean {
 }
 
+export function can_unsubscribe(sub: StreamSubscription): boolean {
+    if (can_unsubscribe_others(sub)) {
```
Might be good to add a comment here explaining that `can_unsubscribe_others` handles checking `can_administer_channel_group` setting.

sahil839 reviewed (commented):
Posted few comments.

sahil839 commented on web/src/stream_ui_updates.ts:
```diff
@@ -218,7 +219,13 @@ export function update_settings_button_for_sub(sub: StreamSubscription): void {
+        } else {
+            $parent.attr("data-tooltip-template-id", "cannot-subscribe-tooltip-template");
+        }
+        const $button_wrapper = $(".settings .stream_settings_header .sub_unsub_button_wrapper");
```
This can still call `initialize_cant_subscribe_popover` once the `data-tooltip-templte-id` attribute is set.

Aditya8840 replied:
Renamed `initialize_cant_subscribe_popover` to `initialize_subscription_toggle_disabled_popover` in the second commit to enable reuse here.

sahil839 commented on zerver/lib/streams.py:
```diff
@@ -1119,6 +1127,37 @@ def bulk_can_remove_subscribers_from_streams(
+        get_recursive_membership_groups(user_profile).values_list("id", flat=True)
+    )
+
+    cannot_administer_atleast_one_stream: bool = True
```
Variable name is fine, but value should also be updated accordingly.

sahil839 reviewed (commented):
Posted some comments.

Aditya8840 committed:
streams: Rename popover for subscription toggle restrictions.

This commit makes the function reusable, helping reduce code
duplication.

timabbott commented:
@Aditya8840 have you addressed the feedback? One should always post a batch of comments indicating that when pushing changes (at least if you've done so -- if it was just a rebase so far and it's not ready for review, you've done the right thing).

Aditya8840 commented on zerver/tests/test_subs.py:
```diff
@@ -4265,43 +4268,168 @@ def test_remove_invalid_user(self) -> None:
-        self.make_stream(stream_name)
+        stream = self.make_stream(stream_name)
+
+        remove_others_group = check_add_user_group(
```
Addressed by using anonymous group with `check_unsubscribing_user ` function.

Aditya8840 commented on api_docs/unmerged.d/ZF-1abcca.md:
```diff
@@ -0,0 +1,9 @@
+  [`PATCH /streams/{stream_id}`](/api/update-stream): Added `can_unsubscribe_group`
+  parameter for setting the user group whose members can unsubscribe themselves from channel.
+* [`DELETE /users/me/subscriptions`](/api/unsubscribe): Restricted unsubscription to
+  organization administrators and members of the channel's `can_unsubscribe_group`.
```
Updated!

```
Unsubscription is allowed for organization administrators, users who can administer the channel or remove other subscribers, and members of the channel's `can_unsubscribe_group`.
```

Aditya8840 commented:
@sahil839, Ready for your review again!

sahil839 commented on zerver/lib/streams.py:
```diff
@@ -1342,6 +1382,10 @@ def list_to_streams(
     existing_streams: list[Stream] = []
     missing_stream_dicts: list[StreamDict] = []
     existing_stream_map = bulk_get_streams(user_profile.realm, stream_set)
+    if unsubscribing_self and not bulk_can_unsubscribe_self_from_streams(
```
Would be good to check how we handle the case when user tries to unsubscribe from an already unsubscribed stream before and after this change? I guess we ignored the already unsubscribed streams that the user could access but not sure what we did for streams which user did not have access to like a private stream.

Aditya8840 replied:
I might have misunderstood your original point, but if you were suggesting that we add tests for the edge cases where users attempt to unsubscribe from streams they’re not subscribed to, I’ve now added tests for both scenarios. Please let me know if you were referring to something else!

Aditya8840 replied:
Now, we raise an `"Insufficient permission"` error if the stream is inaccessible to the user. I've added a test for this at `zerver/tests/test_channel_permissions.py` (lines 859–863) in the same PR.

Do we want to maintain the original behavior instead?

sahil839 reviewed (commented):
Posted some comments.

Aditya8840 committed:
streams: Raise error for invalid channel ID on private streams.

Previously, when a user attempted to unsubscribe others from a
stream they didn't have access to, the system returned an
"Insufficient permission" error. This commit updates the behavior
to raise an "Invalid channel ID" error instead.

Aditya8840 commented:
@zulipbot remove "has conflicts"

Aditya8840 commented:
I've updated the code to raise an "Invalid channel ID" error when the user doesn't have access to a private stream. Added a fourth commit for this — I believe it was missing previously (though I'm not entirely sure).

sahil839 commented on zerver/tests/test_channel_permissions.py:
```diff
@@ -737,6 +737,133 @@ def check_unsubscribing_user(
+
+        # Test that a user in the can_unsubscribe_group can
+        # unsubscribe themselves
+        self.subscribe(aaron, stream.name)
```
I think user can be subscribed in `check_unsubscribing_user`. And then for testing the case where user does not have access to stream, we can skip using `check_unsubscribing_user` and directly call `self.api_delete`.

sahil839 commented on zerver/tests/test_channel_permissions.py:
```diff
@@ -564,6 +564,7 @@ def check_unsubscribing_user(
             user: UserProfile,
             can_remove_subscribers_group: NamedUserGroup | UserGroupMembersData,
             expect_fail: bool = False,
+            user_has_channel_access: bool = True,
```
Rather than having two parameters `expect_fail` and `user_has_channel_access`, this can have a `error_msg` parameter which will be `None` when user has permission.

sahil839 commented on zerver/tests/test_channel_permissions.py:
```diff
@@ -737,6 +737,133 @@ def check_unsubscribing_user(
+
+        # Test that a user is not subscribed to the inaccessible stream.
+        stream = self.make_stream("private_stream", invite_only=True)
+        stream_name_list = [stream.name]
```
Would be good to add user to the `can_unsubscribe_group` group as well, even when the user is not subscribed.

sahil839 reviewed (commented):
Posted few comments. Also tested manually, and we do not disable the UI in some places when user cannot unsubscribe like checkmark in the left panel of stream settings UI and unsubscribe button in user profile modal.

sahil839 reviewed (commented):
There are a couple of things to discuss in #design, but looks good overall.

alya commented on web/templates/stream_settings/browse_streams_list_item.hbs:
```diff
@@ -2,9 +2,9 @@
 
     {{#if subscribed}}
-        <div class="check checked sub_unsub_button">
+        <div class="check checked sub_unsub_button {{#unless should_display_subscription_button}}disabled{{/unless}}">
```
Yeah, our general pattern is to use reduced opacity for disabled buttons, without changing the color.

Aditya8840 replied:
reduced opacity to `0.5` for channels the user can't unsubscribe from

sahil839 reviewed (commented):
Posted one comment. Looks good otherwise.

Aditya8840 commented on web/src/user_profile.ts:
```diff
@@ -283,6 +283,8 @@ export function get_user_unsub_streams(user_id: number): dropdown_widget.Option[
 function format_user_stream_list_item_html(stream: StreamSubscription, user: User): string {
     const show_unsubscribe_button =
         people.can_admin_user(user) || stream_data.can_unsubscribe_others(stream);
+    const unsubscribe_button_disabled =
```
oops, missed that part

Aditya8840 commented:
@timabbott, This PR is ready for your review.

sahil839 commented on web/src/user_profile.ts:
```diff
@@ -282,7 +282,8 @@ export function get_user_unsub_streams(user_id: number): dropdown_widget.Option[
     const show_unsubscribe_button =
-        people.can_admin_user(user) || stream_data.can_unsubscribe_others(stream);
+        stream_data.can_unsubscribe_others(stream) ||
+        (people.is_my_user_id(user.user_id) && stream_data.can_unsubscribe(stream));
```
API docs mention that user can unsubscribe their own bot if they have access to the channel irrespective of any seting. But this is not true here anymore.

Aditya8840 committed:
streams: Add realm-level `can_unsubscribe_group` setting.

Fixes: zulip/zulip/pull/34099#issuecomment-3071553100

Aditya8840 commented on zerver/tests/test_channel_permissions.py:
```diff
@@ -737,6 +737,139 @@ def check_unsubscribing_user(
+            anonymous_group_member_dict,
+            acting_user=aaron,
+        )
+        check_unsubscribing_user(aaron)
```
Removed this line from docs.

Aditya8840 commented:
Added org-level `can_unsubscribe_group` permission setting.

sahil839 commented on zerver/openapi/zulip.yaml:
```diff
@@ -12412,21 +12418,32 @@ paths:
+            by the [`can_remove_subscribers_group`][can-remove-parameter]
+            for the channel.
+
+        **Changes**: Before Zulip 11.0 (feature level ZF-1abcca), anyone who
```
This needs to be changed to `Zulip 12.0`.

Aditya8840 replied:
fixed

sahil839 reviewed (commented):
I reviewed the commit for adding the realm level setting and posted some comments. You would need to fix the tests which are failing due to conflicts in migration.

Aditya8840 commented on zerver/openapi/zulip.yaml:
```diff
@@ -29270,10 +29308,17 @@ components:
             to administer the channel in order to modify this setting.
 
-            **Changes**: New in Zulip 11.0 (feature level ZF-1abcca).
+            **Changes**: Prior to Zulip 11.0 (feature level ZF-9ffedd), Only stream-level
```
Addressed and merged both changelogs into a single one (`ZF-1abcca`).

Aditya8840 commented:
I fixed the CI failure and this PR is ready for review.

alya commented:
Please tag your mentors for review when PRs are ready, to avoid delays.

@sbansal1999 @rsashank FYI

sahil839 commented on web/templates/stream_settings/browse_streams_list_item.hbs:
```diff
@@ -14,6 +14,22 @@
+                <template id="cannot-unsubscribe-from-{{name}}-stream-tooltip-template">
+                    <span>
+                        {{#tr}}
+                            You are not allowed to unsubscribe from this channel.
```
For some reason I am seeing empty tooltips for both streams user can and cannot unsubscribe from
[Screencast from 2025-11-07 18-55-07.webm](https://github.com/user-attachments/assets/216c8d14-9a16-474c-84e7-0267da77915d)

sahil839 commented on web/styles/subscriptions.css:
```diff
@@ -784,10 +784,6 @@ h4.user_group_setting_subsection_title {
                 opacity: 0.5;
             }
 
-            &.checked svg {
```
This should have been removed in the commit that added this.

sahil839 reviewed (commented):
Posted some comments.

alya commented:
Thanks for the work here, @Aditya8840 ! Closing in favor of updated work in #38503.
